returnn/search.py

```diff
@@ -14,6 +14,178 @@
 import recipe.i6_core.util as util
 
 
+class ReturnnSearchJob(Job):
+    """
+    Given a model checkpoint, run search task with RETURNN
+    """
+
+    def __init__(
+        self,
+        search_data,
+        model_checkpoints,
+        epoch,
+        returnn_config,
+        *,
+        log_verbosity=3,
+        device="gpu",
+        time_rqmt=4,
+        mem_rqmt=4,
+        cpu_rqmt=2,
+        returnn_python_exe=None,
+        returnn_root=None,
+    ):
+        """
+        :param dict[str] search_data: dataset used for search
+        :param dict[int, Checkpoint] model_checkpoints: Contains checkpoints pointing to TF models.
```

Comment:
It is not a good idea to pass a list of checkpoints, and then select only one by epoch, as the hash will change when you alter this list, even if the actually selected checkpoint is the same.

So we have to find variants for the inputs that are correct, but also work in any case. My suggestion would be to have the following two options:
 - model output folder + epoch
 - fixed `Checkpoint` object

Reply:
actually this should be of type `dict[int, List[Checkpoint]]`. Also, the checkpoints list is usually fixed and created in `ReturnnTrainingJob`. 
The problem here is that `epoch` can be a `tk.Variable` and so `.get()` should be only called from here. So I guess the first option would work. Actually this was what I had before. But at some point we agreed that it is better to use the `Checkpoint` class.

Reply:
I don't quite understand. 

This should be a single `Checkpoint` object. Not a model output folder + epoch. And also not a dict of multiple checkpoints.

> The problem here is that epoch can be a tk.Variable

We should never do this. I think we can always replace such code. If there are cases where this is not easy/clear, maybe we should discuss separately, as this is sth more fundamental.

Reply:
Yes I would agree that this should be a single checkpoint. This is what I had before. But at some point, we wanted to select recognition epochs using a Job in Sisyphus config. This returns a `tk.Variable` representing the epoch (using `self.output_var`). That's why I do this now. The solution for this would be not to use `Checkpoint` but pass the model output folder + epoch. If we want to ignore this for now, then a single `Checkpoint` will work.

Reply:
> But at some point, we wanted to select recognition epochs using a Job in Sisyphus config. This returns a `tk.Variable` representing the epoch (using `self.output_var`).

I think this should be solved in some way as well (but independent from this PR). Maybe we should open a separate issue to discuss, if the solution is not obvious.

`ReturnnSearchJob` definitely should just accept a single `Checkpoint`.

Maybe we can also extend/modify `Checkpoint` itself, but that again should be a separate PR/discussion, not related to `ReturnnSearchJob`, which just uses `Checkpoint`.

Reply:
Having to pass "epoch" as `tk.Variable` is not necessary, as in this case there can (and should) be a separate job that "selects" a model. The only situation where you need to have this model_folder + epoch input is if you want to work with models that can be deleted during the training (e.g. intermediate checkpoints), and using the `update` functionality of a Job. This is a very specific case, and is usually not needed (You can tell RETURNN to keep some specific epochs I think, I am always fine with the pre-defined 80 and 160 so I never checked this).

Reply:
ok then i will just use single checkpoint

Reply:
> Having to pass "epoch" as `tk.Variable` is not necessary, as in this case there can (and should) be a separate job that "selects" a model.

Yes, this is what I had in mind. Btw, the current `Checkpoint` class, will it actually have the same hash then? This `SelectEpochJob` (or however you would call it) likely would put a symlink or so, right? And `Checkpoint` would use the path of the symlink, which has a different hash than the checkpoint file itself? I don't know. Just asking. If this is a problem, as said, we should discuss in a separate issue, not here.

> The only situation where you need to have this model_folder + epoch input is if you want to work with models that can be deleted during the training (e.g. intermediate checkpoints), and using the `update` functionality of a Job.

I don't understand. Why can't you use `Checkpoint` for this case?

Reply:
> Why can't you use Checkpoint for this case?
It is probably even possible, I never looked at this again. But this is too specific to be considered here anyway.

> Btw, the current Checkpoint class, will it actually have the same hash then

No, this will not be the case. If e.g. epoch 247 would be determined to be the "best" epoch from a Job that does this, and you would start with epoch 247 "by hand", those would be two different recognition jobs. This is of course redundant, but this can not be resolved as this is not possible with the Graph concept defined by Sisyphus. It is not the "Content" (e.g. the physical checkpoint) that is hashed, but the ownership. So Sisyphus would create 2 decoding jobs, because "Training.<hash>/output/checkpoint.247" and "GetBestCheckpoint.<hash>/output/best_checkpoint" are different, even if "best_checkpoint" is just a symlink to checkpoint.247

Reply:
>> Btw, the current Checkpoint class, will it actually have the same hash then
>
> No, this will not be the case.

We could fix this. E.g. Sisyphus could resolve symlinks (but only so far as they keep staying relative to the work dir) before it takes the hash. Or sth like that. They are surely many potential ways to fix this.

Or, maybe actually this is fine for the case of "best checkpoint", because when you rerun this part, and due to non-determinism, the best epoch changes, the "best checkpoint" hash will still be the same as before, which is maybe exactly what we want. Not sure.

But let's further discuss elsewhere, this is all off-topic for this PR here.

Reply:
ok i guess this can be resolved then.